File: app/src/main/java/com/pluscubed/velociraptor/cache/Way.java
package com.pluscubed.velociraptor.cache;

import com.google.auto.value.AutoValue;
import com.pluscubed.velociraptor.api.ApiResponse;
import com.pluscubed.velociraptor.api.osmapi.Coord;
import com.squareup.sqldelight.RowMapper;

import java.util.ArrayList;
import java.util.List;

@AutoValue
public abstract class Way implements wayModel {
    public static final Factory<Way> FACTORY = new wayModel.Factory<>(AutoValue_Way::new);

    public static final RowMapper<Way> SELECT_BY_COORD = FACTORY.select_by_coordMapper();

    public static List<Way> fromResponse(ApiResponse response) {
        List<Way> ways = new ArrayList<>();
        for (int i = 0; i < response.coords.size() - 1; i++) {
            Coord coord1 = response.coords.get(i);
            Coord coord2 = response.coords.get(i + 1);

            double clat = (coord1.lat + coord2.lat) / 2;
            double clon = (coord1.lon + coord2.lon) / 2;

            Way way = new AutoValue_Way(clat, clon,
                    (long) response.speedLimit, response.timestamp,
                    coord1.lat, coord1.lon, coord2.lat, coord2.lon, response.roadName);
            ways.add(way);
        }
        return ways;
    }

    public ApiResponse toResponse() {
        ApiResponse response = new ApiResponse();
        response.speedLimit = (int) maxspeed();
        response.timestamp = timestamp();
        response.roadName = road();
        return response;
    }
}

File: app/src/main/java/com/pluscubed/velociraptor/api/cache/Way.kt
package com.pluscubed.velociraptor.api.cache

import androidx.room.Entity
import com.pluscubed.velociraptor.api.Coord
import com.pluscubed.velociraptor.api.LimitResponse
import java.util.*

@Entity(primaryKeys = ["clat", "clon"])
data class Way(
    val clat: Double,
    val clon: Double,
    val maxspeed: Int,
    val timestamp: Long,
    val lat1: Double,
    val lon1: Double,
    val lat2: Double,
    val lon2: Double,
    val road: String,
    val origin: Int
) {

    companion object {
        fun fromResponse(response: LimitResponse): List<Way> {
            val ways = ArrayList<Way>()
            for (i in 0 until response.coords.size - 1) {
                val coord1 = response.coords[i]
                val coord2 = response.coords[i + 1]

                val clat = (coord1.lat + coord2.lat) / 2
                val clon = (coord1.lon + coord2.lon) / 2

                val way = Way(
                    clat, clon,
                    response.speedLimit,
                    response.timestamp,
                    coord1.lat, coord1.lon, coord2.lat, coord2.lon,
                    response.roadName,
                    response.origin
                )
                ways.add(way)
            }
            return ways
        }
    }

    fun toResponse(): LimitResponse {
        return LimitResponse(
            speedLimit = maxspeed,
            timestamp = timestamp,
            roadName = road,
            origin = origin,
            coords = Arrays.asList(Coord(lat1, lon1), Coord(lat2, lon2))
        )
    }

}